File: AgoraChat-ObjectiveC/AgoraChat/Code/Presence/PresenceManager.m
//
//  PresenceManager.m
//  EaseIM
//
//  Created by lixiaoming on 2022/2/8.
//  Copyright © 2022 lixiaoming. All rights reserved.
//

#import "PresenceManager.h"

static PresenceManager *presenceManager = nil;

@interface PresenceManager()

@end

@implementation PresenceManager
+ (instancetype)sharedInstance
{
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        presenceManager = [[PresenceManager alloc] init];
        [[[AgoraChatClient sharedClient] presenceManager] addDelegate:presenceManager delegateQueue:nil];
        [[AgoraChatClient sharedClient] addDelegate:presenceManager delegateQueue:nil];
    });
    return presenceManager;
}

//- (void)initialize
//{
//    [[[AgoraChatClient sharedClient] presenceManager] fetchSubscribedMembersWithPageNum:0 pageSize:50 Completion:^(NSArray<NSString *> *members, AgoraChatError *error) {
//        self.subscribedMembers = members;
//    }];
//}

- (NSMutableArray<NSString*>*)subscribedMembers
{
    if(!_subscribedMembers) {
        _subscribedMembers = [NSMutableArray<NSString*> array];
    }
    return _subscribedMembers;
}

- (NSMutableDictionary*)presences
{
    if(!_presences) {
        _presences = [NSMutableDictionary dictionary];
    }
    return _presences;
}

+ (NSDictionary*)presenceImages
{
    return @{
        @PRESENCESTATUS_ONLINE:kPresenceOnlineDescription,
        @PRESENCESTATUS_OFFLINE:kPresenceOfflineDescription,
        @PRESENCESTATUS_BUSY:kPresenceBusyDescription,
        @PRESENCESTATUS_DONOTDISTURB:kPresenceDNDDescription,
        @PRESENCESTATUS_LEAVE:kPresenceLeaveDescription,
        @PRESENCESTATUS_CUSTOM:@"custom"
    };
}

+ (NSDictionary*)whiteStrokePresenceImages
{
    return @{
        @PRESENCESTATUS_ONLINE:@"Online_whitestroke",
        @PRESENCESTATUS_OFFLINE:@"Offline_whitestroke",
        @PRESENCESTATUS_BUSY:@"Busy_whitestroke",
        @PRESENCESTATUS_DONOTDISTURB:@"Do Not Disturb_whitestroke",
        @PRESENCESTATUS_LEAVE:@"Away_whitestroke",
        @PRESENCESTATUS_CUSTOM:@"custom_whitestroke"
    };
}

+ (NSDictionary*)showStatus
{
    return @{
        @PRESENCESTATUS_ONLINE:kPresenceOnlineDescription,
        @PRESENCESTATUS_OFFLINE:kPresenceOfflineDescription,
        @PRESENCESTATUS_BUSY:kPresenceBusyDescription,
        @PRESENCESTATUS_DONOTDISTURB:kPresenceDNDDescription,
        @PRESENCESTATUS_LEAVE:kPresenceLeaveDescription
    };
}

- (void) subscribe:(NSArray<NSString*>*)members completion:(void(^)(NSArray<AgoraChatPresence*>* presences,AgoraChatError*error))aCompletion
{
//    NSMutableArray<NSString*>* members = [NSMutableArray array];
//    for(NSString* member in submembers) {
//        if([self.subscribedMembers containsObject:member])
//            continue;
//        [members addObject:member];
//    }
    NSInteger index = 0;
    NSInteger count = members.count;
    while(count > 0) {
        NSRange range;
        range.location = 100*index;
        if(count > 100) {
            range.length = 100;
        }else
            range.length = count;
        count -= range.length;
        NSArray* arr = [members subarrayWithRange:range];
        WEAK_SELF
        [[[AgoraChatClient sharedClient] presenceManager] subscribe:arr expiry:7*24*3600 completion:^(NSArray<AgoraChatPresence *> *presences, AgoraChatError *error) {
            if(!error) {
                [weakSelf.subscribedMembers addObjectsFromArray:arr];
                NSMutableArray<NSString*>* users = [NSMutableArray array];
                for (AgoraChatPresence* presence in presences) {
                    if(presence && presence.publisher.length > 0) {
                        [users addObject:presence.publisher];
                        [weakSelf.presences setObject:presence forKey:presence.publisher];
                    }
                }
                if(presences.count > 0)
                    [[NSNotificationCenter defaultCenter] postNotificationName:PRESENCES_UPDATE  object:users];
            }
            
            if(aCompletion)
                aCompletion(presences,error);
        }];
    }
}

- (void) unsubscribe:(NSArray<NSString*>*) members completion:(void(^)(AgoraChatError*error))aCompletion
{
    NSInteger index = 0;
    NSInteger count = members.count;
    while(members.count > 0) {
        NSRange range;
        range.location = 100*index;
        if(count > 100) {
            range.length = 100;
        }else
            range.length = count;
        NSArray* arr = [members subarrayWithRange:range];
        WEAK_SELF
        [[[AgoraChatClient sharedClient] presenceManager] unsubscribe:arr completion:^(AgoraChatError *error) {
            if(!error) {
                [weakSelf.subscribedMembers removeObjectsInArray:arr];
            }
        }];
    }
}

- (void) fetchPresencesByMembers:(NSArray*)members completion:(void(^)(NSArray<AgoraChatPresence*>* presences,AgoraChatError*error))aCompletion
{
    NSInteger index = 0;
    NSInteger count = members.count;
    while(count > 0) {
        NSRange range;
        range.location = 100*index;
        if(count > 100) {
            range.length = 100;
        }else
            range.length = count;
        count -= range.length;
        NSArray* arr = [members subarrayWithRange:range];
        [[[AgoraChatClient sharedClient] presenceManager] fetchPresenceStatus:arr completion:^(NSArray<AgoraChatPresence *> *presences, AgoraChatError *error) {
            if(!error) {
                NSMutableArray<NSString*>* users = [NSMutableArray array];
                for (AgoraChatPresence* presence in presences) {
                    if(presence && presence.publisher.length > 0) {
                        [users addObject:presence.publisher];
                        [self.presences setObject:presence forKey:presence.publisher];
                    }
                }
                if(presences.count > 0)
                    [[NSNotificationCenter defaultCenter] postNotificationName:PRESENCES_UPDATE  object:users];
            }
            
            if(aCompletion)
                aCompletion(presences,error);
        }];
    }
}
- (void) publishPresenceWithDescription:(NSString*)aDescription completion:(void(^)(AgoraChatError*error))aCompletion
{
    if([aDescription isEqualToString:kPresenceOnlineDescription])
        aDescription = @"";
    [[[AgoraChatClient sharedClient] presenceManager] publishPresenceWithDescription:aDescription completion:^(AgoraChatError *error) {
        if(aCompletion)
            aCompletion(error);
    }];
}

- (void) presenceStatusDidChanged:(NSArray<AgoraChatPresence*>*)presences
{
    NSMutableArray<NSString*>* users = [NSMutableArray array];
    for (AgoraChatPresence* presence in presences) {
        if(presence && presence.publisher.length > 0) {
            [users addObject:presence.publisher];
            [self.presences setObject:presence forKey:presence.publisher];
        }
    }
    if(presences.count > 0)
        [[NSNotificationCenter defaultCenter] postNotificationName:PRESENCES_UPDATE  object:users];
}

- (void)connectionStateDidChange:(AgoraChatConnectionState)aConnectionState
{
    AgoraChatPresence* presence = [self.presences objectForKey:[AgoraChatClient sharedClient].currentUsername];
    if(aConnectionState == AgoraChatConnectionDisconnected) {
        if(presence) {
            for (AgoraChatPresenceStatusDetail* detail in presence.statusDetails) {
                detail.status = 0;
            }
            [[NSNotificationCenter defaultCenter] postNotificationName:PRESENCES_UPDATE  object:@[[AgoraChatClient sharedClient].currentUsername]];
        }
    }
}

+ (NSInteger)fetchStatus:(AgoraChatPresence*)presence
{
    if(!presence)
        return 0;
    if(presence.statusDetails.count <= 0)
        return 0;
    NSInteger retStatus= 0;
    for(AgoraChatPresenceStatusDetail* detail in presence.statusDetails) {
        if(detail.status == 1){
            retStatus = 1;
            break;
        }
    }
    if(retStatus != 0) {
        if(presence.statusDescription.length > 0) {
            retStatus = PRESENCESTATUS_CUSTOM;
            if([presence.statusDescription isEqualToString:kPresenceOnlineDescription])
                retStatus = PRESENCESTATUS_ONLINE;
            if([presence.statusDescription isEqualToString:kPresenceBusyDescription])
                retStatus = PRESENCESTATUS_BUSY;
            if([presence.statusDescription isEqualToString:kPresenceDNDDescription])
                retStatus = PRESENCESTATUS_DONOTDISTURB;
            if([presence.statusDescription isEqualToString:kPresenceLeaveDescription])
                retStatus = PRESENCESTATUS_LEAVE;
        }
    }
    return retStatus;
}

+ (NSString*)formatOfflineStatus:(NSUInteger)lasttime
{
    NSDate* date = [NSDate date];
    NSUInteger secodes = [date timeIntervalSince1970] - lasttime;
    NSString* timeStr = @"";
    if(secodes < 60)
        timeStr = [NSString stringWithFormat:@"%lus",secodes];//unit seconds
    else if(secodes < 60*60)
        timeStr = [NSString stringWithFormat:@"%lum",secodes/60];//unit minutes
    else if(secodes < 24*60*60)
        timeStr = [NSString stringWithFormat:@"%luh",secodes/60/60];//unit hours
    else if(secodes < 7*24*60*60)
        timeStr = [NSString stringWithFormat:@"%lud",secodes/60/60/24];//unit hours
    else
        timeStr = [NSString stringWithFormat:@"%luw",secodes/60/60/24/7];//unit hours
    return [NSString stringWithFormat:@"Online %@ ago",timeStr];
}

@end

File: AgoraChat-Swift/AgoraChat-Swift/code/Presence/PresenceManager.swift
//
//  PresenceManager.swift
//  AgoraChat-Swift
//
//  Created by 冯钊 on 2022/10/26.
//

import UIKit

class PresenceManager: NSObject {
    enum State: Int {
        case offline = 0
        case online = 1
        case busy = 100
        case doNotDisturb = 101
        case leave = 102
        case custom = 103
    }
    
    static let presenceImagesMap: [State: String] = [
        .offline: "Offline",
        .online: "Online",
        .busy: "Busy",
        .doNotDisturb: "Do not Disturb",
        .leave: "Leave",
        .custom: "custom"
    ]
    
    static let whiteStrokePresenceImagesMap: [State: String] = [
        .offline: "Offline_whitestroke",
        .online: "Online_whitestroke",
        .busy: "Busy_whitestroke",
        .doNotDisturb: "Do not Disturb_whitestroke",
        .leave: "Leave_whitestroke",
        .custom: "custom_whitestroke"
    ]
    
    static let showStatusMap: [State: String] = [
        .offline: "Offline".localized,
        .online: "Online".localized,
        .busy: "Busy".localized,
        .doNotDisturb: "Do Not Disturb".localized,
        .leave: "Leave".localized
    ]
    
    class func formatOfflineTimespace(presence: AgoraChatPresence?) -> String {
        guard let lastTime = presence?.lastTime else {
            return self.showStatusMap[.offline]!
        }
        let timeinterval = Int(Date().timeIntervalSince1970 - Double(lastTime))
        let timeStr: String
        if timeinterval < 60 {
            timeStr = "1 \("Minute".localized)"
        } else if timeinterval < 60 * 60 {
            timeStr = "\(timeinterval / 60) \("Minutes".localized)"
        } else if timeinterval < 60 * 60 * 24 {
            timeStr = "\(timeinterval / 60 / 60) \("Hours".localized)"
        } else if timeinterval < 60 * 60 * 24 * 7 {
            timeStr = "\(timeinterval / 60 / 60 / 24) \("Days".localized)"
        } else {
            timeStr = "\(timeinterval / 60 / 60 / 24 / 7) \("Weeks".localized)"
        }
        return String(format: "Online %@ ago".localized, timeStr)
    }
    
    static let shared = PresenceManager()
    
    class func fetchStatus(presence: AgoraChatPresence?) -> State {
        guard let presence = presence, let statusDetails = presence.statusDetails else {
            return .offline
        }
        if statusDetails.count <= 0 {
            return .offline
        }
        var state: State = .offline
        for i in statusDetails {
            if i.status == 1 {
                state = .online
                break
            }
        }
        
        if state != .offline, let statusDescription = presence.statusDescription, statusDescription.count > 0 {
            state = .custom
            for i in self.showStatusMap where statusDescription == i.value {
                state = i.key
                break
            }
        }
        return state
    }
    
    private var subscribedMembers: [String] = []
    var presences: [String: AgoraChatPresence] = [:]
    
    private override init() {
        super.init()
        AgoraChatClient.shared().presenceManager?.add(self, delegateQueue: nil)
        AgoraChatClient.shared().add(self, delegateQueue: nil)
    }
    
    func subscribe(members: [String], completion: ((_ presence: [AgoraChatPresence]?, _ error: AgoraChatError?) -> Void)?) {
        var index = 0
        var count = members.count
        while count > 0 {
            var range = NSRange(location: index * 100, length: 0)
            if count > 100 {
                range.length = 100
                index += 1
            } else {
                range.length = count
            }
            count -= range.length
            let array = Array(members[range.location..<range.location + range.length])
            AgoraChatClient.shared().presenceManager?.subscribe(array, expiry: 7 * 24 * 3600, completion: { presences, error in
                if let presences = presences {
                    self.subscribedMembers.append(contentsOf: array)
                    var users: [String] = []
                    for presence in presences {
                        if presence.publisher.count > 0 {
                            users.append(presence.publisher)
                            self.presences[presence.publisher] = presence
                        }
                    }
                    if presences.count > 0 {
                        DispatchQueue.main.async {
                            NotificationCenter.default.post(name: PresenceUpdateNotification, object: users)
                        }
                    }
                }
                DispatchQueue.main.async {
                    completion?(presences, error)
                }
            })
        }
    }

    func unsubscribe(members: [String], completion: ((_ error: AgoraChatError?) -> Void)?) {
        var index = 0
        var count = members.count
        while count > 0 {
            var range = NSRange(location: index * 100, length: 0)
            if count > 100 {
                range.length = 100
                index += 1
            } else {
                range.length = count
            }
            count -= range.length
            let array = Array(members[range.location..<range.location + range.length])
            AgoraChatClient.shared().presenceManager?.unsubscribe(array, completion: { error in
                if error == nil {
                    self.subscribedMembers.remove(elements: array)
                }
                completion?(error)
            })
        }
    }

    func publishPresence(description: String?, completion: ((_ error: AgoraChatError?) -> Void)?) {
        var description = description
        if description == PresenceManager.showStatusMap[.online] {
            description = nil
        }
        AgoraChatClient.shared().presenceManager?.publishPresence(withDescription: description, completion: { error in
            completion?(error)
        })
    }
}

extension PresenceManager: AgoraChatPresenceManagerDelegate {
    func presenceStatusDidChanged(_ presences: [AgoraChatPresence]) {
        var users: [String] = []
        for presence in presences {
            if presence.publisher.count > 0 {
                users.append(presence.publisher)
                self.presences[presence.publisher] = presence
            }
        }
        if presences.count > 0 {
            NotificationCenter.default.post(name: PresenceUpdateNotification, object: users)
        }
    }
}

extension PresenceManager: AgoraChatClientDelegate {
    func connectionStateDidChange(_ aConnectionState: AgoraChatConnectionState) {
        guard let currentUsername = AgoraChatClient.shared().currentUsername, let presence = self.presences[currentUsername], let statusDetails = presence.statusDetails else {
            return
        }
        if aConnectionState == .disconnected {
            for detail in statusDetails {
                detail.status = PresenceManager.State.offline.rawValue
            }
            NotificationCenter.default.post(name: PresenceUpdateNotification, object: [currentUsername])
        }
    }
}
